### DISCUSSION TITLE
How to filter, flatten and rewrite REST-api results

### DISCUSSION THREAD

**Discussion by ansibleguy76** [ansibleguy76/ansibleforms#126] [Q&A] [answered]
A question I often hear, is when you get a detailed and deep-nested rest result, how can you cleanup, filter and flatten rest result.
The answer is JQ.  JQ stands for Json Query and is a powerfull language to manipulate JSON data and the REST api results are commonly JSON.

Every REST function (fn.fnRestBasic, fn.fnRestJwt, ...) has a JQ property that will run against the result.

https://ansibleforms.com/expressions/#rest-api-with-basic-authentication
https://ansibleforms.com/expressions/#builtin-jq-functions

Point your questions about jq and rest result filtering here.  Or better, share your snippets.

**Comment by MalfuncEddie** [comment]
```
{
  "id": 419,
  "url": "",
  "display": "10.29.0.0/24",
  "family": {
    "value": 4,
    "label": "IPv4"
  },
  "prefix": "10.29.0.0/24",
  "site": null,
  "vrf": null,
  "tenant": {
    "id": 118,
    "url": "",
    "display": "",
    "name": "",
    "slug": ""
  },
  "vlan": null,
  "status": {
    "value": "active",
    "label": "Active"
  },
  "role": null,
  "is_pool": false,
  "mark_utilized": false,
  "description": "",
  "comments": "",
  "tags": [],
  "custom_fields": {
    "gateway": "",
    "segment_name": "THIS NAME AS COLUM",
    "uuid": "",
    "Sync_source": "",
    "Managed_by": ""
  }
}
```

how do I get custom_fields.segment_name as columns (enum)
```

- name: Prefix
    type: enum
    expression: fn.fnRestAdvanced( 'get',
      '','',{},
      '[.results[]] ')
    columns:
      - display
      - custom_fields.segment_name
```

**Reply by ansibleguy76** [reply]
it is not possible this way.  maybe jq can do this.  otherwise a local javascript expression can surely convert this.

**Comment by ansibleguy76** [comment]
From your sample, can you show me what the output should be ?

**Reply by MalfuncEddie** [reply]
In the form I can show "root" values like "id", "display" and that shows a dropdown of all the id or displays.

When  is set a colum like "custom_fields.segment_name" it does not work and it just show the intire object in the dropdown field

**Reply by ansibleguy76** [reply, marked as answer]
you need to rewrite the output with jq.

you already use [.records[]] => you must enhance this.

[jqplay snippet](https://jqplay.org/s/luAvoMSJCvr)

`[.records[] | {id:.id,display:.display,segment:.custom_fields.segment_name}]`
Try this snippet, you will see

**Reply by MalfuncEddie** [reply]
tnx this is working !
